pkg/cloudprovider/aws/instance.go

```diff
@@ -321,26 +324,45 @@ func (p *InstanceProvider) getCapacityType(constraints *v1alpha1.Constraints, in
 	return v1alpha1.CapacityTypeOnDemand
 }
 
-// filterInstanceTypes is used to eliminate GPU instance types from the list of possible instance types when a
-// non-GPU instance type will work.  If the list of instance types consists of both GPU and non-GPU types, then only
-// the non-GPU types will be returned.  If it has only GPU types, the list will be returned unaltered.
+// filterInstanceTypes is used to eliminate less desirable instance types (like GPUs) from the list of possible instance types when
+// a set of more appropriate instance types would work. If a set of more desirable instance types is not found, then the original slice
+// of instance types are returned.
 func (p *InstanceProvider) filterInstanceTypes(instanceTypes []cloudprovider.InstanceType) []cloudprovider.InstanceType {
 	var genericInstanceTypes []cloudprovider.InstanceType
 	for _, it := range instanceTypes {
+		// if the cluster is ipv6, then use nitro instance types since xen does not support ipv6 prefix allocations
+		if p.ipv6DNS != nil && *it.(*InstanceType).Hypervisor != "nitro" {
+			continue
+		}
 		itRes := it.Resources()
-		if resources.IsZero(itRes[v1alpha1.ResourceAWSNeuron]) &&
-			resources.IsZero(itRes[v1alpha1.ResourceAMDGPU]) &&
-			resources.IsZero(itRes[v1alpha1.ResourceNVIDIAGPU]) {
-			genericInstanceTypes = append(genericInstanceTypes, it)
+		if !resources.IsZero(itRes[v1alpha1.ResourceAWSNeuron]) ||
+			!resources.IsZero(itRes[v1alpha1.ResourceAMDGPU]) ||
+			!resources.IsZero(itRes[v1alpha1.ResourceNVIDIAGPU]) {
+			continue
 		}
+		genericInstanceTypes = append(genericInstanceTypes, it)
 	}
-	// if we got some subset of non-GPU types, then prefer to use those
+	// if we got some subset of instance types, then prefer to use those
 	if len(genericInstanceTypes) != 0 {
 		return genericInstanceTypes
 	}
 	return instanceTypes
 }
 
+// injectIPv6DNS adds the ipv6DNS to a copied constraint's KubeletConfiguration.ClusterDNS if a ClusterDNS was not already provided
+func (p *InstanceProvider) injectIPv6DNS(constraints *v1alpha1.Constraints) *v1alpha1.Constraints {
```

Comment:
Why are we using mutation? Can we have a common method where we retrieve this value?

Reply:
It has been mutated to not mutate :)